diff --git a/include/aws/mqtt/private/client_impl.h b/include/aws/mqtt/private/client_impl.h
index 4dc13cf4aac8c14a1c6fac0aa1cabb43e6ad0605..6361bbb7380c370884bcd4a9d2467dd4fd8cca14 100644
--- a/include/aws/mqtt/private/client_impl.h
+++ b/include/aws/mqtt/private/client_impl.h
@@ -153,6 +153,25 @@ struct aws_mqtt_reconnect_task {
     struct aws_allocator *allocator;
 };
 
+struct request_timeout_wrapper;
+
+/* used for timeout task */
+struct request_timeout_task_arg {
+    uint16_t packet_id;
+    struct aws_mqtt_client_connection_311_impl *connection;
+    struct request_timeout_wrapper *task_arg_wrapper;
+};
+
+/*
+ * We want the timeout task to be able to destroy the forward reference from the operation's task arg structure
+ * to the timeout task.  But the operation task arg structures don't have any data structure in common.  So to allow
+ * the timeout to refer back to a zero-able forward pointer, we wrap a pointer to the timeout task and embed it
+ * in every operation's task arg that needs to create a timeout.
+ */
+struct request_timeout_wrapper {
+    struct request_timeout_task_arg *timeout_task_arg;
+};
+
 /* The lifetime of this struct is from subscribe -> suback */
 struct subscribe_task_arg {
 
@@ -172,6 +191,8 @@ struct subscribe_task_arg {
         aws_mqtt_suback_fn *single;
     } on_suback;
     void *on_suback_ud;
+
+    struct request_timeout_wrapper timeout_wrapper;
 };
 
 /* The lifetime of this struct is the same as the lifetime of the subscription */
diff --git a/source/client.c b/source/client.c
index c332c4da86425f96e7aca2b0a5da79071b479fd3..8c138bfac2748e3973043e514bb18c7d0ef05e71 100644
--- a/source/client.c
+++ b/source/client.c
@@ -93,25 +93,6 @@ void mqtt_connection_set_state(
     connection->synced_data.state = state;
 }
 
-struct request_timeout_wrapper;
-
-/* used for timeout task */
-struct request_timeout_task_arg {
-    uint16_t packet_id;
-    struct aws_mqtt_client_connection_311_impl *connection;
-    struct request_timeout_wrapper *task_arg_wrapper;
-};
-
-/*
- * We want the timeout task to be able to destroy the forward reference from the operation's task arg structure
- * to the timeout task.  But the operation task arg structures don't have any data structure in common.  So to allow
- * the timeout to refer back to a zero-able forward pointer, we wrap a pointer to the timeout task and embed it
- * in every operation's task arg that needs to create a timeout.
- */
-struct request_timeout_wrapper {
-    struct request_timeout_task_arg *timeout_task_arg;
-};
-
 static void s_request_timeout(struct aws_channel_task *channel_task, void *arg, enum aws_task_status status) {
     (void)channel_task;
     struct request_timeout_task_arg *timeout_task_arg = arg;
@@ -1894,6 +1875,22 @@ static enum aws_mqtt_client_request_state s_subscribe_send(uint16_t packet_id, b
         aws_mem_release(message->allocator, message);
     }
 
+    if (task_arg->connection->operation_timeout_ns != UINT64_MAX) {
+        /* TODO: timing should start from the message written into the socket, which is aws_io_message->on_completion
+         * invoked, but there are bugs in the websocket handler (and maybe also the h1 handler?) where we don't properly
+         * fire the on_completion callbacks. */
+        struct request_timeout_task_arg *timeout_task_arg =
+            s_schedule_timeout_task(task_arg->connection, packet_id);
+        if (timeout_task_arg) {
+            /*
+             * Set up mutual references between the operation task args and the timeout task args.  Whoever runs first
+             * "wins", does its logic, and then breaks the connection between the two.
+             */
+            task_arg->timeout_wrapper.timeout_task_arg = timeout_task_arg;
+            timeout_task_arg->task_arg_wrapper = &task_arg->timeout_wrapper;
+        }
+    }
+
     if (!task_arg->tree_updated) {
         aws_mqtt_topic_tree_transaction_commit(&task_arg->connection->thread_data.subscriptions, &transaction);
         task_arg->tree_updated = true;
@@ -1959,6 +1956,18 @@ static void s_subscribe_complete(
             error_code,
             task_arg->on_suback_ud);
     }
+
+    /*
+     * If we have a forward pointer to a timeout task, then that means the timeout task has not run yet.  So we should
+     * follow it and zero out the back pointer to us, because we're going away now.  The timeout task will run later
+     * and be harmless (even vs. future operations with the same packet id) because it only cancels if it has a back
+     * pointer.
+     */
+    if (task_arg->timeout_wrapper.timeout_task_arg) {
+        task_arg->timeout_wrapper.timeout_task_arg->task_arg_wrapper = NULL;
+        task_arg->timeout_wrapper.timeout_task_arg = NULL;
+    }
+
     for (size_t i = 0; i < list_len; i++) {
         aws_array_list_get_at(&task_arg->topics, &topic, i);
         s_task_topic_release(topic);
@@ -2129,6 +2138,18 @@ static void s_subscribe_single_complete(
             error_code,
             task_arg->on_suback_ud);
     }
+
+    /*
+     * If we have a forward pointer to a timeout task, then that means the timeout task has not run yet.  So we should
+     * follow it and zero out the back pointer to us, because we're going away now.  The timeout task will run later
+     * and be harmless (even vs. future operations with the same packet id) because it only cancels if it has a back
+     * pointer.
+     */
+    if (task_arg->timeout_wrapper.timeout_task_arg) {
+        task_arg->timeout_wrapper.timeout_task_arg->task_arg_wrapper = NULL;
+        task_arg->timeout_wrapper.timeout_task_arg = NULL;
+    }
+
     s_task_topic_release(topic);
     aws_array_list_clean_up(&task_arg->topics);
     aws_mqtt_packet_subscribe_clean_up(&task_arg->subscribe);
@@ -2353,6 +2374,22 @@ static enum aws_mqtt_client_request_state s_resubscribe_send(
         aws_mem_release(message->allocator, message);
     }
 
+    if (task_arg->connection->operation_timeout_ns != UINT64_MAX) {
+        /* TODO: timing should start from the message written into the socket, which is aws_io_message->on_completion
+         * invoked, but there are bugs in the websocket handler (and maybe also the h1 handler?) where we don't properly
+         * fire the on_completion callbacks. */
+        struct request_timeout_task_arg *timeout_task_arg =
+            s_schedule_timeout_task(task_arg->connection, packet_id);
+        if (timeout_task_arg) {
+            /*
+             * Set up mutual references between the operation task args and the timeout task args.  Whoever runs first
+             * "wins", does its logic, and then breaks the connection between the two.
+             */
+            task_arg->timeout_wrapper.timeout_task_arg = timeout_task_arg;
+            timeout_task_arg->task_arg_wrapper = &task_arg->timeout_wrapper;
+        }
+    }
+
     return AWS_MQTT_CLIENT_REQUEST_ONGOING;
 
 handle_error:
@@ -2416,6 +2453,17 @@ static void s_resubscribe_complete(
 
 clean_up:
 
+    /*
+     * If we have a forward pointer to a timeout task, then that means the timeout task has not run yet.  So we should
+     * follow it and zero out the back pointer to us, because we're going away now.  The timeout task will run later
+     * and be harmless (even vs. future operations with the same packet id) because it only cancels if it has a back
+     * pointer.
+     */
+    if (task_arg->timeout_wrapper.timeout_task_arg) {
+        task_arg->timeout_wrapper.timeout_task_arg->task_arg_wrapper = NULL;
+        task_arg->timeout_wrapper.timeout_task_arg = NULL;
+    }
+
     /* We need to cleanup the subscribe_task_topics, since they are not inserted into the topic tree by resubscribe. We
      * take the ownership to clean it up */
     for (size_t i = 0; i < list_len; i++) {
diff --git a/tests/CMakeLists.txt b/tests/CMakeLists.txt
index 3494aa2eaed3585aa54d2dc494b7543f3d37b569..659b20466d0b8c990f5b212690bdf4e0fb342920 100644
--- a/tests/CMakeLists.txt
+++ b/tests/CMakeLists.txt
@@ -79,6 +79,7 @@ add_test_case(mqtt_clean_session_discard_previous)
 add_test_case(mqtt_clean_session_keep_next_session)
 add_test_case(mqtt_connection_publish_QoS1_timeout)
 add_test_case(mqtt_connection_unsub_timeout)
+add_test_case(mqtt_connection_sub_timeout)
 add_test_case(mqtt_connection_publish_QoS1_timeout_connection_lost_reset_time)
 add_test_case(mqtt_connection_ping_norm)
 add_test_case(mqtt_connection_ping_no)
diff --git a/tests/v3/connection_state_test.c b/tests/v3/connection_state_test.c
index 6a3e086fd459d70636b508f0410f67bcf8b8be04..8ed877aa2d5f1747e76d188a45b82a6ad04fd41c 100644
--- a/tests/v3/connection_state_test.c
+++ b/tests/v3/connection_state_test.c
@@ -3069,6 +3069,60 @@ AWS_TEST_CASE_FIXTURE(
     s_clean_up_mqtt_server_fn,
     &test_data)
 
+/**
+ * Test that a subscribe completes with a timeout when the broker does not send SUBACK.
+ */
+static int s_test_mqtt_connection_sub_timeout_fn(struct aws_allocator *allocator, void *ctx) {
+    (void)allocator;
+    struct mqtt_connection_state_test *state_test_data = ctx;
+
+    struct aws_mqtt_connection_options connection_options = {
+        .user_data = state_test_data,
+        .clean_session = false,
+        .client_id = aws_byte_cursor_from_c_str("client1234"),
+        .host_name = aws_byte_cursor_from_c_str(state_test_data->endpoint.address),
+        .socket_options = &state_test_data->socket_options,
+        .on_connection_complete = s_on_connection_complete_fn,
+        .ping_timeout_ms = DEFAULT_TEST_PING_TIMEOUT_MS,
+        .protocol_operation_timeout_ms = 3000,
+        .keep_alive_time_secs = 16960, /* basically stop automatically sending PINGREQ */
+    };
+
+    ASSERT_SUCCESS(aws_mqtt_client_connection_connect(state_test_data->mqtt_connection, &connection_options));
+    s_wait_for_connection_to_complete(state_test_data);
+
+    /* Disable automatic acknowledgements so the server never sends SUBACK. */
+    mqtt_mock_server_disable_auto_ack(state_test_data->mock_server);
+
+    struct aws_byte_cursor topic = aws_byte_cursor_from_c_str("/test/topic");
+    uint16_t sub_packet_id = aws_mqtt_client_connection_subscribe(
+        state_test_data->mqtt_connection,
+        &topic,
+        AWS_MQTT_QOS_AT_LEAST_ONCE,
+        s_on_publish_received,
+        state_test_data,
+        NULL,
+        s_on_suback,
+        state_test_data);
+    ASSERT_TRUE(sub_packet_id > 0);
+
+    s_wait_for_subscribe_to_complete(state_test_data);
+    ASSERT_UINT_EQUALS(state_test_data->subscribe_complete_error, AWS_ERROR_MQTT_TIMEOUT);
+
+    ASSERT_SUCCESS(
+        aws_mqtt_client_connection_disconnect(state_test_data->mqtt_connection, s_on_disconnect_fn, state_test_data));
+    s_wait_for_disconnect_to_complete(state_test_data);
+
+    return AWS_OP_SUCCESS;
+}
+
+AWS_TEST_CASE_FIXTURE(
+    mqtt_connection_sub_timeout,
+    s_setup_mqtt_server_fn,
+    s_test_mqtt_connection_sub_timeout_fn,
+    s_clean_up_mqtt_server_fn,
+    &test_data)
+
 /**
  * Test that connection is healthy, user set the timeout for request, and connection lost will reset timeout.
  */
